# <center> Variables, expressions, and statements <center>

In this file we will see:
1. Values 
2. Variables and types
3. Variable names and keywords
4. Operators and operands
5. Expressions and statements
6. Order of operations
7. Exercises

## 1. Values
A **value** is one of the basic things a program works with, like a letter or a number.
For instance:

In [1]:
"Hola Mundo"

'Hola Mundo'

In [2]:
2.0

2.0

## 2. Variables and types
A **variable** is a name that refers to a value. In Python, there are four types of variables: 
* <font color=red>Int</font>: A type value that represents integer numbers.
* <font color=red>Floating - point</font>: A type of value that represents numbers with a decimal part.
* <font color=red>String</font>: Represents a sequence of characters.
* <font color=red>Bool</font>: Represents two types of values: <code>true</code> and <code>false</code>.


In [3]:
#Integer number
1

1

In [4]:
# Float number
1.13

1.13

In [5]:
#String
"Hola Mundo"

'Hola Mundo'

In [6]:
#Bool
False

False

## 2.1 The <code>type</code> fucntion

In Python, there is a pre-build function that can identify the type of a variable and it is called <code>type</code>.

In [3]:
#Type of an integer
type(1)

int

In [4]:
#Type of a float
type(3.14)

float

In [5]:
#Type of a String
type("Hola,Mundo!")

str

In [6]:
#Type of a boolean
type(False)

bool

## 2.2 Variable assignment
A **variable assignment** creates a new variable and gives value to it.

In [11]:
message = "Welcome to class!"
print(message)

Welcome to class!


In [12]:
score = 9
score

9

In [13]:
division = 14.7
division


14.7

## 2.3 Type conversion

Sometimes we want to convert some type of variable to another. To do that, we can use the pre-built functions that Python provides:
* <code>int(var)</code>
* <code>float(var)</code>
* <code>str(var)</code>
* <code>bool(var)</code>

In [11]:
#Convert from string to integer
num = "3"

In [15]:
int(num)

3

In [16]:
#Convert from float to str
score = 8.7

In [17]:
str(score)

'8.7'

## 3. Variable names and keywords
* It is recommended to use variables that are **meaningful**.
* Variable names can be arbitrarily long and can contain letters and numbers but **must begin with a letter**. 
* There are certain reserved keywords that cannot be used as a variable name.

**Here are some examples of an incorrect name for a variable:**

In [14]:
a = 1
b = 3.14
xy = "Student"
xy@ = 23

SyntaxError: invalid syntax (<ipython-input-14-31baa723534d>, line 4)

**Now we will see some examples of a correct variable name:**

In [15]:
finalScore = 10
final_score = 8
car_type = "Honda"
average = 8.8

<table>
    <tr >
        <th style=background-color:grey; colspan="5"><center>Reserved Keywords</center></th>
    </tr>
    <tr style=background-color:#33e4ff;>
        <td style=text-align:left;>and</td>
        <td style=text-align:left;>del</td>
        <td style=text-align:left;>from</td>
        <td style=text-align:left;>not</td>
        <td style=text-align:left;>while</td>
    </tr>
    <tr style=background-color:#33abff;>
        <td style=text-align:left;>as</td>
        <td style=text-align:left;>elif</td>
        <td style=text-align:left;>global</td>
        <td style=text-align:left;>or</td>
        <td style=text-align:left;>with</td>
    </tr>
    <tr style=background-color:#33e4ff;>
        <td style=text-align:left;>assert</td>
        <td style=text-align:left;>else</td>
        <td style=text-align:left;>if</td>
        <td style=text-align:left;>pass</td>
        <td style=text-align:left;>yield</td>
    </tr>
    <tr style=background-color:#33abff;>
        <td style=text-align:left;>break</td>
        <td style=text-align:left;>except</td>
        <td style=text-align:left;>import</td>
        <td style=text-align:left;>print</td>
        <td style=text-align:left;></td>
    </tr>
    <tr style=background-color:#33e4ff;>
        <td style=text-align:left;>class</td>
        <td style=text-align:left;>exec</td>
        <td style=text-align:left;>in</td>
        <td style=text-align:left;>raise</td>
        <td></td>
    </tr>
    <tr style=background-color:#33abff;>
        <td style=text-align:left;>continue</td>
        <td style=text-align:left;>finally</td>
        <td style=text-align:left;>is</td>
        <td style=text-align:left;>return</td>
        <td></td>
    </tr>
    <tr style=background-color:#33e4ff;>
        <td style=text-align:left;>def</td>
        <td style=text-align:left;>for</td>
        <td style=text-align:left;>lambda</td>
        <td style=text-align:left;>try</td>
        <td></td>
    </tr>
</table>

## 4. Operators and operands

* **Operators** are special symbols that represent computations like addition and multiplication. 
* The values the operator is applied to are called **operands**.
* The operators $+$, $-$, $*$, $/$, and $**$ perform addition, subtraction, multiplication, division, and exponentiation.

**Example 1. Adding two grades.**

In [16]:
grade1 = 8 
grade2 = 5

In [17]:
grade1 + grade2

13

**Example 2. Given hour and minutes, convert them to minutes.**

In [18]:
hour = 10
minutes = 43

In [19]:
hour * 60 + minutes

643

**Example 3. Given the minutes, convert to hours.**

In [20]:
minutes = 77

In [21]:
minutes / 60

1.2833333333333334

**Example 4. Assigning students to care for cats.**

In [22]:
students = 8
cats = 3

In [23]:
students // cats

2

**Example 5. Determine if a number is even.**

In [24]:
number = 8

In [25]:
number % 2

0

**Example 6. Raise $a$ to the power of $b$.**

In [26]:
a = 2
b = 5

In [27]:
a ** b

32

----

## 5. Expressions and statements

* An **expression** is a combination of values, variables, and operators.
    * Examples: 
    1. $x$,
    2. $17$,
    3. $x + 17$
    
    
* A **statement** is a unite of code that Python can execute.
    * Examples: 
    1. $x = 23$,
    2. <code>print(x)</code>

----

## 6. Order of operations

* When more than one operator appears in an expression, the order of evaluation depends on the **rules of precedence**. 
* Python follows mathematical convention:
    1. Parentheses.
    2. Exponentiation.
    3. Multiplication and Division.
    4. Addition and subtraction.

**Example 1. Use of parentheses**

In [28]:
2 * (3 - 1) 

4

In [29]:
2 * 3 - 1 

5

**Notice that** $$2 * (3-1) \neq 2 * 3 - 1$$
This is because the way to solve the left side is as follows: 
$$ 2 * (3 - 1)$$
$$ = 2 * (2)$$
$$= 4$$

And the way to solve the right side is as follows:

$$2 * 3 - 1$$
$$= 6 - 1$$
$$= 5$$.


**Example 2. Exponentiation before multiplication and division**

In [30]:
2 * 4 ** 2

32

In [31]:
(2 * 4) ** 2

64

**Notice that** $$2 * 4 ^{2} \neq (2 * 4)^{2} $$
This is because the way to solve the left side is as follows: 
$$ 2 * 4 ^{2}$$
$$ = 2 * 16$$
$$= 32$$

And the way to solve the right side is as follows:

$$(2 * 4)^{2}$$
$$=(8)^{2}$$
$$= 64$$

**Example 3. Same hierarchy of operations**

If we have operators with the same hierarchy in an expression, then we perform the operation from left to right.

In [32]:
minutes * hour / 60

12.833333333333334

**Example 4. Other uses of parentheses**

Sometimes we can use parentheses to make a statement clearer:

In [33]:
celsius = 28
(celsius * 9/5) + 32

82.4

----

## 7. Exercises

**1. Compute the area of a triangle of base 10 cm and height 15 cm.** 

$$TriangleArea =\frac{base * height}{2}$$

**2. Compute the area of a circle of radius 5.**

$$CircleArea = \pi * r^{2}$$

**3. Compute the average of three scores: $8.5$, $7.2$, and $9.3$**

**4. Given a Celsius degree, convert it to Fahrenheit.**

$$ F = \frac{9}{5} * C + 32$$

**5. A group of friends go out for a dinner and the total bill to pay is of 87.50 USD, if they would like to give a tip of 15%. How much the tip should be?**


**EXTRA EXERCISE**
<br>*Ask the user for an amount of mexican pesos so it can be converted to dollars. Consider that $19.92 MXN is equivalent to 1 USD*